addons/stock/report/report_stock_reception.py

```diff
@@ -0,0 +1,285 @@
+        for product_id, outs in products_to_outs.items():
+            for out in outs:
+                # only show pickings or MOs, no SOs
+                source = out.picking_id if out.picking_id else out._get_source_document()
+                if not source:
+                    continue
+
+                qty_to_reserve = out.product_qty
+                product_uom = out.product_id.uom_id
+                if 'done' not in picking_states and out.state == 'partially_available':
+                    qty_to_reserve -= out.product_uom._compute_quantity(out.reserved_availability, product_uom)
+                moves_in = self.env['stock.move']
+                qty_done = 0
+                for move_in_qty, move_in in product_to_qty_to_assign[out.product_id]:
+                    moves_in |= move_in
+                    if float_compare(qty_done + move_in_qty, qty_to_reserve, precision_rounding=product_uom.rounding) <= 0:
+                        qty_to_add = move_in_qty
+                        move_in_qty = 0
+                    else:
+                        qty_to_add = qty_to_reserve - qty_done
+                        move_in_qty -= qty_to_add
+                    qty_done += qty_to_add
+                    if move_in_qty:
+                        product_to_qty_to_assign[out.product_id][0] = (move_in_qty, move_in)
+                    else:
+                        product_to_qty_to_assign[out.product_id] = product_to_qty_to_assign[out.product_id][1:]
+                    if float_compare(qty_to_reserve, qty_done, precision_rounding=product_uom.rounding) == 0:
+                        break
+
+                if not float_is_zero(qty_done, precision_rounding=product_uom.rounding):
+                    sources_to_lines[source].append(self._prepare_report_line(
+                        qty_done, product_id, out, source, move_ins=moves_in))
+
+                # draft qtys can be shown but not assigned
+                qty_expected = product_to_qty_draft.get(product_id, 0)
+                if float_compare(qty_to_reserve, qty_done, precision_rounding=product_uom.rounding) > 0 and\
+                        not float_is_zero(qty_expected, precision_rounding=product_uom.rounding):
+                    to_expect = min(qty_expected, qty_to_reserve - qty_done)
+                    sources_to_lines[source].append(self._prepare_report_line(to_expect, product_id, out, source, is_qty_assignable=False))
+                    product_to_qty_draft[product_id] -= to_expect
+
+        # show already assigned moves
+        for product_id, ins_to_qtys in product_to_total_assigned.items():
+            total_assigned = 0
+            all_out_ids = []
+            in_ids = set()
+            for (in_move, qty_assigned) in ins_to_qtys:
+                in_ids.add(in_move.id)
+                total_assigned += qty_assigned
+                all_out_ids += in_move.move_dest_ids.ids
+            out_moves = self.env['stock.move'].browse(list(set(all_out_ids)))
```

Comment:
Careful with `set`, it is not a ordered collection. Then order of `out_moves` is "random", then the call of method can return two order different for the same request. (we can use a OrderedSet to keep the order)

Reply:
set removed via your better `product_to_total_assigned` suggestion